### PR TITLE
C++/WinRT composable authoring support 

### PR DESCRIPTION
**Pull request by kennykerr**
This update brings xlang's C++ projection to parity with C++/WinRT with regard to composable authoring support. I've manually validated against the internal C++/WinRT unit tests, but not yet ported over all of the tests. That's coming up soon. 

This update also includes numerous other bug fixes flushed out through testing, but I still need to go through and complete test coverage. There are some known issues, but overall the C++ tool is now far more stable and feature complete.

### CODE REVIEW

File: src/test/cpp/compare/out/winrt/Component.Composable.h [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        int32_t WINRT_CALL CreateInstance(void* baseInterface, void** innerInterface, void** value) noexcept final
+        {
+            try
+            {
+                if (innerInterface) *innerInterface = nullptr;
+                Windows::Foundation::IInspectable winrt_impl_innerInterface;
+                *value = nullptr;
+                typename D::abi_guard guard(this->shim());
+                *value = detach_from<Component::Composable::Base>(this->shim().CreateInstance(*reinterpret_cast<Windows::Foundation::IInspectable const*>(&baseInterface), winrt_impl_innerInterface));
+            if (innerInterface) *innerInterface = detach_abi(winrt_impl_innerInterface);
```
Comment (DefaultRyan) [inline src/test/cpp/compare/out/winrt/Component.Composable.h]: Formatting

File: src/test/cpp/component.idl [hunk 2]
```diff
@@ -161,4 +164,19 @@ namespace Component
             static Windows.Foundation.Collections.IMap<String, Int32> Map();
         }
     }
+
+    namespace Composable
```
Comment (DefaultRyan) [inline src/test/cpp/component.idl]: This needs to preserve the generation of [overridable] interfaces and methods if we're validating parity with C++/WinRT.

File: src/tool/cpp/helpers.h [hunk 3]
```diff
@@ -287,47 +291,56 @@ namespace xlang
...
+            get_interfaces_impl(w, result, info.defaulted, info.overridable, base, info.generic_param_stack, info.type.InterfaceImpl());
             result[name] = std::move(info);
         }
     };
 
     auto get_interfaces(writer& w, TypeDef const& type)
     {
         std::map<std::string, interface_info> result;
-        get_interfaces_impl(w, result, false, false, false, type.InterfaceImpl());
+        get_interfaces_impl(w, result, false, false, false, {}, type.InterfaceImpl());
```
Comment (DefaultRyan) [inline src/tool/cpp/helpers.h]: All these bool flags make the callsites hostile to human readers. Replace the bools with bool enums to preserve readability and prevent accidental mix-and-matching.

Comment (kennykerr) [reply src/tool/cpp/helpers.h]: Agreed, I've been meaning to clean this up as it has evolved into something more complex.

File: src/test/cpp/component.idl (hunk 2 above)
Comment (kennykerr) [reply src/test/cpp/component.idl]: Agreed, I need to flesh out the testing. Going to do that in a separate branch and PR. I'm also looking into cmake code coverage in tandem to ensure we cover all our supported scenarios and code paths.

File: src/test/cpp/compare/out/winrt/Component.Composable.h (hunk 1 above)
Comment (kennykerr) [reply src/test/cpp/compare/out/winrt/Component.Composable.h]: Yep, I need to do a formatting sweep. So glad we have the generated output to visually inspect. 😉

**Pull request merged**